modules/core/image/src/Plugin/ImageEffect/AutoOrientImageEffect.php

```diff
@@ -0,0 +1,204 @@
+   *   A logger instance.
+   * @param \Drupal\Core\File\FileSystemInterface $file_system
+   *   The file system service.
+   */
+  public function __construct(array $configuration, string $plugin_id, array $plugin_definition, LoggerInterface $logger, FileSystemInterface $file_system) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $logger);
+    $this->fileSystem = $file_system;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('logger.factory')->get('image'),
+      $container->get('file_system'),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applyEffect(ImageInterface $image) {
+
+    // Only proceed if the exif PHP extension is loaded.
+    if (!extension_loaded('exif')) {
+      return TRUE;
+    }
+
+    // Only work with supported image types.
+    $mime_types = [
+      'image/jpeg',
+      'image/png',
+      'image/tiff',
+    ];
+    if (!in_array($image->getMimeType(), $mime_types)) {
+      return TRUE;
+    }
+
+    // Load EXIF orientation.
+    $orientation = $this->loadExifOrientation($image->getSource());
+
+    // If an orientation is not specified, bail.
+    if (is_null($orientation)) {
+      return TRUE;
+    }
+
+    // Apply transformations based on the EXIF Orientation value.
+    // http://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/EXIF.html:
+    // 1 = Horizontal (normal)                 [top-left].
+    // 2 = Mirror horizontal                   [top-right].
+    // 3 = Rotate 180                          [bottom-right].
+    // 4 = Mirror vertical                     [bottom-left].
+    // 5 = Mirror horizontal and rotate 270 CW [left-top].
+    // 6 = Rotate 90 CW                        [right-top].
+    // 7 = Mirror horizontal and rotate 90 CW  [right-bottom].
+    // 8 = Rotate 270 CW                       [left-bottom].
+    switch ($orientation) {
+      case 2:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        break;
+
+      case 3:
+        $result = $image->apply('rotate', ['degrees' => 180]);
+        break;
+
+      case 4:
+        $result = $image->apply('mirror', ['y_axis' => TRUE]);
+        break;
+
+      case 5:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        if ($result) {
+          $result = $image->apply('rotate', ['degrees' => 270]);
+        }
+        break;
+
+      case 6:
+        $result = $image->apply('rotate', ['degrees' => 90]);
+        break;
+
+      case 7:
+        $result = $image->apply('mirror', ['x_axis' => TRUE]);
+        if ($result) {
+          $result = $image->apply('rotate', ['degrees' => 90]);
+        }
+        break;
+
+      case 8:
+        $result = $image->apply('rotate', ['degrees' => 270]);
+        break;
+    }
```

Comment:
For larger chunks of code that appear directly derived from [other code](https://git.drupalcode.org/project/image_effects/-/blob/730a1e6f5f947ef4d6eec5d359c6968ce93594f9/src/Plugin/ImageToolkit/Operation/gd/AutoOrient.php#L45-86), my understanding is that there should be some sort of inline attribution indicating where the code was copied/modified from.

We can argue about whether this code is too much of boiler-plate for the GPL attribution/copyright-listing requirements to come into play (obligatory IAANAL), but at the very least I'd argue for perma-linking to the source of the copied code since that also has come in handy to me in the past when troubleshooting errors that crept into adapted code.

Reply:
Agreed! I also added a comment to `transformDimensions()` because that was also inspired by Image Effects code.

Commit: https://github.com/farmOS/farmOS/pull/941/commits/29cb9cac312364f35c0bef84c7b5799d289223de